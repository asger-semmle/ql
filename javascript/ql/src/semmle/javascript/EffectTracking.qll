/**
 * Provides classes and predicates for reasoning about the ordering of side effects
 * on a value.
 */
import javascript

module EffectTracking {
  private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps

  abstract class Configuration extends string {
    bindingset[this]
    Configuration() { any() }

    /**
     * Holds if the operation at the given location has an effect on the given value,
     * in a way that matters for this effect tracking configuration.
     */
    abstract predicate isTrackedSideEffect(ControlFlowNode loc, DataFlow::SourceNode value);

    /**
     * Holds if a function call itself should be marked as a side effect if the
     * callee has side effects.
     */
    predicate shouldSummarizeCalls() { any() }

    /**
     * Holds if the operation at the given location has an effect on the given value,
     * or (transitively) calls a function that performs side effects on the value.
     */
    predicate hasSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      isTrackedSideEffect(loc, value) or
      calleeHasSideEffects(DataFlow::valueNode(loc), value)
    }

    /**
     * Holds if the given function has a side effect that manipulates `value`.
     */
    predicate functionHasSideEffects(Function f, DataFlow::SourceNode value) {
      exists (ControlFlowNode loc |
        hasSideEffect(loc, value) and
        loc.getContainer() = f)
    }

    /**
     * Holds if the function invoked by `invoke` performs side effects on the given value
     * passed into the call.
     */
    predicate calleeHasSideEffects(DataFlow::InvokeNode invoke, DataFlow::SourceNode value) {
      shouldSummarizeCalls() and
      (
        // Passed as argument
        exists (DataFlow::Node arg, Function target, Parameter param |
          FlowSteps::argumentPassing(invoke, arg, target, param) and
          value.flowsTo(arg) and
          functionHasSideEffects(target, DataFlow::parameterNode(param)))
        or
        // Captured by nested function
        exists (Function target |
          target = invoke.getACallee() and
          functionHasSideEffects(target, value))
      )
    }

    /**
     * Holds if the given basic block has a effect on the given value.
     */
    private predicate basicBlockHasSideEffects(ReachableBasicBlock block, DataFlow::SourceNode value) {
      exists (ControlFlowNode loc | 
        loc.getBasicBlock() = block and
        hasSideEffect(loc, value))
    }

    /**
     * Gets the last side effect to reach `loc`, unless `loc` is an internal node in a basic block
     * without any side effects.
     *
     * In the latter case, the last effect should be obtained using `getSummarizedSideEffectInBasicBlock`.
     */
    private ControlFlowNode getLastSideEffectAtInternalNode(ControlFlowNode loc, DataFlow::SourceNode value) {
      if hasSideEffect(loc, value) then
        result = loc
      else if isStartOfBasicBlock(loc) then
        exists (ControlFlowNode pred | pred = loc.getAPredecessor() |
          if basicBlockHasSideEffects(pred.getBasicBlock(), value) then
            result = getLastSideEffectAtInternalNode(pred, value)
          else
            result = getLastSideEffectAtInternalNode(pred.getBasicBlock().getFirstNode(), value))
      else if basicBlockHasSideEffects(loc.getBasicBlock(), value) then
        result = getLastSideEffectAtInternalNode(loc.getAPredecessor(), value)
      else
        none() // don't store the recent effect on internal nodes
    }

    /**
     * Gets the last side effect to reach the given basic block, assuming the block has no side effects of its own.
     */
    private ControlFlowNode getSummarizedSideEffectInBasicBlock(BasicBlock block, DataFlow::SourceNode value) {
      result = getLastSideEffectAtInternalNode(block.getFirstNode(), value)
      and not basicBlockHasSideEffects(block, value)
    }

    /**
     * Gets the last side effect to affect the given value at the given location.
     */
    pragma[inline]
    ControlFlowNode getLastSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      result = getLastSideEffectAtInternalNode(loc, value)
      or
      result = getSummarizedSideEffectInBasicBlock(loc.getBasicBlock(), value)
    }

    /**
     * Gets the last operation to affect the given value before the given function returns.
     */
    ControlFlowNode getLastSideEffectInFunction(Function f, DataFlow::SourceNode value) {
      result = getLastSideEffect(f.getExit(), value)
    }

    /**
     * Gets the last operation to affect the given value inside the callee of `invoke`.
     *
     * `value` refers to the value passed to the call, and `calleeAlias` refers to its local alias in the callee.
     */
    ControlFlowNode getLastSideEffectInCallee(DataFlow::InvokeNode invoke, DataFlow::SourceNode value, DataFlow::SourceNode calleeAlias) {
      shouldSummarizeCalls() and
      (
        // Passed as argument
        exists (Function target, Parameter parm, DataFlow::Node arg |
          FlowSteps::argumentPassing(invoke, arg, target, parm) and
          value.flowsTo(arg) and
          calleeAlias = DataFlow::parameterNode(parm) and
          result = getLastSideEffectInFunction(target, calleeAlias))
        or
        // Returned from function
        exists (Function target |
          target = invoke.getACallee() and
          value = invoke and
          calleeAlias = target.getAReturnedExpr().flow().getALocalSource() and
          result = getLastSideEffectInFunction(target, calleeAlias))
        or
        // Captured by nested function
        exists (Function target |
          target = invoke.getACallee() and
          value.getContainer() != target and
          result = getLastSideEffectInFunction(target, value) and
          calleeAlias = value)
      )
    }
  }

  private predicate isStartOfBasicBlock(ControlFlowNode node) {
    exists (BasicBlock b | b.getFirstNode() = node)
  }
}
