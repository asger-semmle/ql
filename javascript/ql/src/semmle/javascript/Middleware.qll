/**
 * Provides generic concepts for modelling the middleware stack of a web application.
 */
import javascript
import semmle.javascript.EffectTracking
import semmle.javascript.frameworks.ExpressMiddleware // imported from here to avoid empty recursion error

module Middleware {
  /**
   * An effect-tracking configuration, tracking installations of route handlers on a router.
   */
  class SetupTracking extends EffectTracking::Configuration {
    SetupTracking() { this = "MiddlewareSetup" }
  
    override predicate isTrackedSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      exists (Setup setup |
        loc = setup.getControlFlowNode() and
        value = setup.getRouter())
    }
  }

  /**
   * A node in a web application's middleware stack model.
   *
   * A node very loosely represents a thing that can handle requests.
   * Incoming requests come from the parent and will be routed to the children,
   * in order of first to last.
   */
  abstract class Node extends DataFlow::Node {
    final ControlFlowNode getControlFlowNode() {
      result = getAstNode()
    }

    /**
     * Gets the local source of the router whose middleware stack is modeled by this node.
     *
     * For routers, this refers to the node itself.
     *
     * For setups that imperatively install route handlers on an existing router, this
     * refers to the router that is being modified.
     */
    abstract DataFlow::SourceNode getRouter();

    /**
     * Gets the last child of this node.
     */
    abstract Node getLastChild();

    /**
     * Gets the previous sibling of this node, that is, the router handler executing
     * before this one.
     *
     * In general there can be multiple previous siblings, forming a sibling graph,
     * but in practice there is rarely more than once.
     */
    abstract Node getPreviousSibling();

    /**
     * Gets the next sibling of this node, that is, the router handler executing
     * after this one.
     *
     * In general there can be multiple next siblings, forming a sibling graph,
     * but in practice there is rarely more than once.
     */
    final Node getNextSibling() {
      this = result.getPreviousSibling()
    }

    /**
     * Gets a child of this node, that is, a router handler that is executed
     * as part of this route handler.
     */
    final Node getAChild() {
      result = this.getLastChild().getPreviousSibling*()
    }
    
    /**
     * Gets the parent of this node, that is, the route handler that dispatches
     * requests to this one.
     */
    final Node getParent() {
      this = result.getAChild()
    }

    /**
     * Gets the path matched by this route handler, relative to its parent.
     *
     * A route handler will reject a request not matching its path.
     *
     * Defaults to the empty string, meaning any request is accepted.
     */
    string getPath() {
      result = ""
    }

    /**
     * Normalizes a path so multiple paths can be joined by concatenation.
     *
     * In particular, this ensures we don't get duplicate `//` path separators.
     */
    bindingset[str]
    private string normalizePath(string str) {
      if str = "/" or str = "" then
        result = ""
      else if str = any(string s) + "/" then
        result = str
      else
        result = str + "/"
    }

    /**
     * Gets the path handled by this node, relative to its parent.
     */
    private string getRelativePath() {
      result = normalizePath(this.getPath())
      or
      not exists(this.getPath()) and
      result = ""
    }

    /**
     * Gets the path handled by this node, relative to the root node.
     */
    final string getAbsolutePath() {
      isTree() and
      (
        result = getParent().getAbsolutePath() + getRelativePath()
        or
        not exists(getParent()) and
        result = getRelativePath()
      )
      or
      // If there are multiple parents, use our best approximation: the absolute paths of any of the parents
      not isTree() and
      result = getParent().getAbsolutePath()
    }

    /**
     * Gets the request method matched by this route handler, or `*` if all
     * request methods are matched.
     *
     * The route handler will reject requests whose request method does not match.
     *
     * Defaults to `*`.
     */
    string getRequestMethod() {
      result = "*"
    }

    /**
     * Gets the request method matched by this router handler and its parents.
     */
    final string getAbsoluteRequestMethod() {
      isRoot() and result = getRequestMethod()
      or
      result = intersectRequestMethodFilters(getParent().getRequestMethod(), getRequestMethod())
    }

    /**
     * Holds if this has no parent.
     */
    final predicate isRoot() {
      not exists(getParent())
    }

    /**
     * Holds if this is part of a tree-shaped middleware stack, which is the common case.
     */
    final predicate isTree() {
      isRoot()
      or
      strictcount(getParent()) = 1 and
      getParent().isTree()
    }

    /**
     * Holds if this handler is preceded by `guard` in the middleware stack.
     *
     * That is, any request seen by this handler is known to have gone through `guard` first.
     */
    final predicate isGuardedBy(Node guard) {
      getParent*().getPreviousSibling+().getAChild*() = guard and
      getAbsolutePath() = guard.getAbsolutePath() + any(string s) and
      exists(intersectRequestMethodFilters(getAbsoluteRequestMethod(), guard.getAbsoluteRequestMethod()))
    }
  }

  /**
   * Gets the intersection of two request method filters.
   */
  bindingset[method1, method2]
  private string intersectRequestMethodFilters(string method1, string method2) {
    method1 = method2 and result = method1
    or
    method1 = "*" and result = method2
    or
    method2 = "*" and result = method1
  }

  /**
   * A middleware node with an indexed list of children.
   *
   * For example, a route setup `app.get("/", handler1, handler2, ...)` has a list of
   * children derived from its argument list.
   */
  private abstract class SequenceNode extends Node {
    /**
     * Gets the `n`th middleware function installed at this point.
     */
    abstract DataFlow::Node getMiddleware(int n);

    /**
     * Gets any of the middleware functions installed at this point.
     */
    final DataFlow::Node getAMiddleware() { result = getMiddleware(_) }

    final int getNumMiddleware() {
      result = count(getAMiddleware())
    }

    final DataFlow::Node getLastMiddleware() {
      result = getMiddleware(getNumMiddleware() - 1)
    }

    override final Node getLastChild() {
      result = getMiddleware(getNumMiddleware() - 1)
    }
  }

  /**
   * An argument in a sequence node.
   */
  private class SequenceElement extends Node {
    SequenceNode parent;
    int index;

    SequenceElement() {
      this = parent.getMiddleware(index)
    }

    override DataFlow::SourceNode getRouter() {
      result = parent.getRouter()
    }
    
    override Node getPreviousSibling() {
      result = parent.getMiddleware(index - 1)
    }
    
    override Node getLastChild() {
      result != this and
      (
        result.(DataFlow::SourceNode).flowsTo(this)
        or
        result.(DataFlow::TrackedNode).flowsTo(this)
      )
    }
  }

  /**
   * An installation of one or more route handlers on a router.
   *
   * The router handler arguments are the children of this node.
   */
  abstract class Setup extends SequenceNode {
    override final Node getPreviousSibling() {
      result.getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(getControlFlowNode().getAPredecessor(), getRouter())
    }
  }

  /**
   * A router, that is, a collection of route handlers with associated paths and request methods.
   */
  class Router extends Node {
    Router() {
      this = any(Node node).getRouter()
    }

    override DataFlow::SourceNode getRouter() {
      result = this
    }
    
    override Node getPreviousSibling() {
      none()
    }
    
    override Node getLastChild() {
      result.(Setup).getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(this.getContainer().getExit(), this)
    }
  }

  /**
   * A functionally defined router, as opposed to routers built up imperatively by adding router handlers on them.
   *
   * For example, an array of route handlers can in some frameworks be used as a route handler itself.
   * In this case, an array literal `[r1, r2]` would be considered a route combinator, with `r1` and `r2` as
   * its children.
   *
   * A route combinator is a `Setup` with themselves as the router.
   */
  abstract class RouteCombinator extends SequenceNode {
    override DataFlow::SourceNode getRouter() {
      result = this
    }

    override Middleware::Node getPreviousSibling() {
      none()
    }
  }

  /**
   * An array literal passed to a route setup.
   */
  private class ArrayRouteCombinator extends Middleware::RouteCombinator, DataFlow::ArrayCreationNode {
    ArrayRouteCombinator() {
      this.flowsTo(any(Node node))
    }
  
    override DataFlow::Node getMiddleware(int n) {
      result = getElement(n)
    }
  }

  /**
   * A call to a function that contains middleware installations.
   */
  class SubSetup extends Middleware::Node, DataFlow::InvokeNode {
    DataFlow::SourceNode router;

    SubSetup() {
      any(SetupTracking tr).calleeHasSideEffects(this, router)
    }

    override DataFlow::SourceNode getRouter() {
      result = router
    }

    override Middleware::Node getLastChild() {
      exists (DataFlow::SourceNode subrouter, ControlFlowNode loc |
        loc = any(SetupTracking tr).getLastSideEffectInCallee(this, router, subrouter) and
        if subrouter = router then
          result.getControlFlowNode() = loc
        else
          result = subrouter)
    }

    override Middleware::Node getPreviousSibling() {
      result.getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(getControlFlowNode().getAPredecessor(), router)
    }
  }
}
