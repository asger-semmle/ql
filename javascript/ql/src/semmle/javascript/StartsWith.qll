/**
 * Provides classes and predicates for recognizing `startsWith` calls and their equivalents.
 */
import javascript

/**
 * A expression that is equivalent to `A.startsWith(B)` or `!A.startsWith(B)`.
 */
abstract class StartsWithCheck extends DataFlow::Node {
  /**
   * Gets the `A` in `A.startsWith(B)`.
   */
  abstract DataFlow::Node getBaseString();

  /**
   * Gets the `B` in `A.startsWith(B)`.
   */
  abstract DataFlow::Node getSubstring();

  /**
   * Gets the polarity if the check.
   *
   * If the polarity is `false` the check returns `true` if the string does not start
   * with the given substring.
   */
  boolean getPolarity() { result = true }
}

/**
 * An expression of form `A.startsWith(B)`.
 */
private class NativeStartsWith extends StartsWithCheck, DataFlow::MethodCallNode {
  NativeStartsWith() {
    getMethodName() = "startsWith" and
    getNumArgument() = 1
  }

  override DataFlow::Node getBaseString() {
    result = getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(0)
  }
}

/**
 * An expression of form `A.indexOf(B) === 0`.
 */
private class IndexOfStartsWith extends StartsWithCheck, DataFlow::ValueNode {
  override EqualityTest astNode;
  DataFlow::MethodCallNode indexOf;

  IndexOfStartsWith() {
    indexOf.getMethodName() = "indexOf" and
    indexOf.getNumArgument() = 1 and
    indexOf.flowsToExpr(astNode.getAnOperand()) and
    astNode.getAnOperand().getIntValue() = 0
  }

  override DataFlow::Node getBaseString() {
    result = indexOf.getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = indexOf.getArgument(0)
  }

  override boolean getPolarity() {
    result = astNode.getPolarity()
  }
}

/**
 * An expression of form `A.indexOf(B)` coerced to a boolean.
 *
 * This is equivalent to `!A.startsWith(B)` since all return values other than zero map to `true`.
 */
private class IndexOfCoercionStartsWith extends StartsWithCheck, DataFlow::MethodCallNode {
  IndexOfCoercionStartsWith() {
    getMethodName() = "indexOf" and
    getNumArgument() = 1 and
    this.flowsToExpr(any(ConditionGuardNode guard).getTest()) // check for boolean coercion
  }

  override DataFlow::Node getBaseString() {
    result = getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(0)
  }

  override boolean getPolarity() {
    result = false
  }
}

/**
 * A call of form `_.startsWith(A, B)` or `ramda.startsWith(A, B)`.
 */
private class LibraryStartsWith extends StartsWithCheck, DataFlow::CallNode {
  LibraryStartsWith() {
    getNumArgument() = 2 and
    exists (DataFlow::SourceNode callee | this = callee.getACall() |
      callee = LodashUnderscore::member("startsWith") or
      callee = DataFlow::moduleMember("ramda", "startsWith")
    )
  }

  override DataFlow::Node getBaseString() {
    result = getArgument(0)
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(1)
  }
}

/**
 * A comparison of form `x[0] === "k"` for some single-character constant `k`.
 */
private class FirstCharacterCheck extends StartsWithCheck, DataFlow::ValueNode {
  override EqualityTest astNode;
  DataFlow::PropRead read;
  Expr constant;
  
  FirstCharacterCheck() {
    read.flowsTo(astNode.getAnOperand().flow()) and
    read.getPropertyNameExpr().getIntValue() = 0 and
    constant.getStringValue().length() = 1 and
    astNode.getAnOperand() = constant
  }

  override DataFlow::Node getBaseString() {
    result = read.getBase()
  }

  override DataFlow::Node getSubstring() {
    result = constant.flow()
  }

  override boolean getPolarity() {
    result = astNode.getPolarity()
  }
}
