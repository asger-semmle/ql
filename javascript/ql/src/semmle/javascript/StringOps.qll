/**
 * Provides classes and predicates for reasoning about string-manipulating expressions.
 */
import javascript

module StringConcatenation {
  /** Gets a data flow node referring to the result of the given concatenation. */
  private DataFlow::Node getAssignAddResult(AssignAddExpr expr) {
    result = expr.flow()
    or
    exists (SsaExplicitDefinition def | def.getDef() = expr |
      result = DataFlow::valueNode(def.getVariable().getAUse()))
  }

  /** Gets the `n`th operand to the string concatenation defining `node`. */
  DataFlow::Node getOperand(DataFlow::Node node, int n) {
    exists (AddExpr add | node = add.flow() |
      n = 0 and result = add.getLeftOperand().flow()
      or
      n = 1 and result = add.getRightOperand().flow())
    or
    exists (TemplateLiteral template | node = template.flow() |
      result = template.getElement(n).flow() and
      not exists (TaggedTemplateExpr tag | template = tag.getTemplate()))
    or
    exists (AssignAddExpr assign | node = getAssignAddResult(assign) |
      n = 0 and result = assign.getLhs().flow()
      or
      n = 1 and result = assign.getRhs().flow())
    or
    exists (DataFlow::ArrayCreationNode array, DataFlow::MethodCallNode call |
      call = array.getAMethodCall("join") and
      call.getArgument(0).mayHaveStringValue("") and
      (
        // step from array element to array
        result = array.getElement(n) and
        node = array
        or
        // step from array to join call
        node = call and
        result = array and
        n = 0
      ))
  }

  /** Gets an operand to the string concatenation defining `node`. */
  DataFlow::Node getAnOperand(DataFlow::Node node) {
    result = getOperand(node, _)
  }

  /** Gets the number of operands to the given concatenation. */
  int getNumOperand(DataFlow::Node node) {
    result = strictcount(getAnOperand(node))
  }

  /** Gets the first operand to the string concatenation defining `node`. */
  DataFlow::Node getFirstOperand(DataFlow::Node node) {
    result = getOperand(node, 0)
  }

  /** Gets the last operand to the string concatenation defining `node`. */
  DataFlow::Node getLastOperand(DataFlow::Node node) {
    result = getOperand(node, getNumOperand(node) - 1)
  }
  
  /**
   * Holds if `src` flows to `dst` through the `n`th operand of the given concatenation operator.
   */
  predicate taintStep(DataFlow::Node src, DataFlow::Node dst, DataFlow::Node operator, int n) {
    src = getOperand(dst, n) and
    operator = dst
  }

  /**
   * Holds if there is a taint step from `src` to `dst` through string concatenation.
   */
  predicate taintStep(DataFlow::Node src, DataFlow::Node dst) {
    taintStep(src, dst, _, _)
  }
}

/**
 * A expression that is equivalent to `A.startsWith(B)` or `!A.startsWith(B)`.
 */
abstract class StartsWithCheck extends DataFlow::Node {
  /**
   * Gets the `A` in `A.startsWith(B)`.
   */
  abstract DataFlow::Node getBaseString();

  /**
   * Gets the `B` in `A.startsWith(B)`.
   */
  abstract DataFlow::Node getSubstring();

  /**
   * Gets the polarity if the check.
   *
   * If the polarity is `false` the check returns `true` if the string does not start
   * with the given substring.
   */
  boolean getPolarity() { result = true }
}

/**
 * An expression of form `A.startsWith(B)`.
 */
private class NativeStartsWith extends StartsWithCheck, DataFlow::MethodCallNode {
  NativeStartsWith() {
    getMethodName() = "startsWith" and
    getNumArgument() = 1
  }

  override DataFlow::Node getBaseString() {
    result = getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(0)
  }
}

/**
 * An expression of form `A.indexOf(B) === 0`.
 */
private class IndexOfStartsWith extends StartsWithCheck, DataFlow::ValueNode {
  override EqualityTest astNode;
  DataFlow::MethodCallNode indexOf;

  IndexOfStartsWith() {
    indexOf.getMethodName() = "indexOf" and
    indexOf.getNumArgument() = 1 and
    indexOf.flowsToExpr(astNode.getAnOperand()) and
    astNode.getAnOperand().getIntValue() = 0
  }

  override DataFlow::Node getBaseString() {
    result = indexOf.getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = indexOf.getArgument(0)
  }

  override boolean getPolarity() {
    result = astNode.getPolarity()
  }
}

/**
 * An expression of form `A.indexOf(B)` coerced to a boolean.
 *
 * This is equivalent to `!A.startsWith(B)` since all return values other than zero map to `true`.
 */
private class IndexOfCoercionStartsWith extends StartsWithCheck, DataFlow::MethodCallNode {
  IndexOfCoercionStartsWith() {
    getMethodName() = "indexOf" and
    getNumArgument() = 1 and
    this.flowsToExpr(any(ConditionGuardNode guard).getTest()) // check for boolean coercion
  }

  override DataFlow::Node getBaseString() {
    result = getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(0)
  }

  override boolean getPolarity() {
    result = false
  }
}

/**
 * A call of form `_.startsWith(A, B)` or `ramda.startsWith(A, B)`.
 */
private class LibraryStartsWith extends StartsWithCheck, DataFlow::CallNode {
  LibraryStartsWith() {
    getNumArgument() = 2 and
    exists (DataFlow::SourceNode callee | this = callee.getACall() |
      callee = LodashUnderscore::member("startsWith") or
      callee = DataFlow::moduleMember("ramda", "startsWith")
    )
  }

  override DataFlow::Node getBaseString() {
    result = getArgument(0)
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(1)
  }
}

/**
 * A comparison of form `x[0] === "k"` for some single-character constant `k`.
 */
private class FirstCharacterCheck extends StartsWithCheck, DataFlow::ValueNode {
  override EqualityTest astNode;
  DataFlow::PropRead read;
  Expr constant;
  
  FirstCharacterCheck() {
    read.flowsTo(astNode.getAnOperand().flow()) and
    read.getPropertyNameExpr().getIntValue() = 0 and
    constant.getStringValue().length() = 1 and
    astNode.getAnOperand() = constant
  }

  override DataFlow::Node getBaseString() {
    result = read.getBase()
  }

  override DataFlow::Node getSubstring() {
    result = constant.flow()
  }

  override boolean getPolarity() {
    result = astNode.getPolarity()
  }
}

/**
 * A expression that is equivalent to `A.includes(B)` or `!A.includes(B)`.
 */
abstract class StringContainsCheck extends DataFlow::Node {
  /** Gets the `A` in `A.includes(B)`. */
  abstract DataFlow::Node getBaseString();

  /** Gets the `B` in `A.includes(B)`. */
  abstract DataFlow::Node getSubstring();

  /**
   * Gets the polarity if the check.
   *
   * If the polarity is `false` the check returns `true` if the string does not start
   * with the given substring.
   */
  boolean getPolarity() { result = true }
}

/**
 * A call to a method named `includes`, assumed to refer to `String.prototype.includes`.
 */
private class NativeStringIncludesCheck extends StringContainsCheck, DataFlow::MethodCallNode {
  NativeStringIncludesCheck() {
    getMethodName() = "includes" and
    getNumArgument() = 1
  }

  override DataFlow::Node getBaseString() {
    result = getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(0)
  }
}

/**
 * A call to `_.includes`, assumed to operate on strings.
 */
private class LibraryStringIncludesCheck extends StringContainsCheck, DataFlow::CallNode {
  LibraryStringIncludesCheck() {
    this = LodashUnderscore::member("includes").getACall()
  }

  override DataFlow::Node getBaseString() {
    result = getArgument(0)
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(1)
  }
}

/**
 * A check of form `A.indexOf(B) !== -1` or similar.
 */
private class IndexOfEqualsContainsCheck extends StringContainsCheck, DataFlow::ValueNode {
  MethodCallExpr indexOf;
  override EqualityTest astNode;

  IndexOfEqualsContainsCheck() {
    exists (Expr index | astNode.hasOperands(indexOf, index) |
      // one operand is of the form `whitelist.indexOf(x)`
      indexOf.getMethodName() = "indexOf" and
      // and the other one is -1
      index.getIntValue() = -1
    )
  }

  override DataFlow::Node getBaseString() {
    result = indexOf.getReceiver().flow()
  }

  override DataFlow::Node getSubstring() {
    result = indexOf.getArgument(0).flow()
  }

  override boolean getPolarity() {
    result = astNode.getPolarity().booleanNot()
  }
}

/**
 * A check of form `A.indexOf(B) >= 0` or similar.
 */
private class IndexOfRelationalContainsCheck extends StringContainsCheck, DataFlow::ValueNode {
  MethodCallExpr indexOf;
  override RelationalComparison astNode;
  boolean polarity;

  IndexOfRelationalContainsCheck() {
    exists (Expr lesser, Expr greater |
      astNode.getLesserOperand() = lesser and
      astNode.getGreaterOperand() = greater and
      indexOf.getMethodName() = "indexOf" and
      indexOf.getNumArgument() = 0 |
      polarity = true and
      greater = indexOf and
      (
        lesser.getIntValue() >= 0
        or
        lesser.getIntValue() = -1 and not astNode.isInclusive()
      )
      or 
      polarity = false and
      lesser = indexOf and
      (
        greater.getIntValue() = -1
        or
        greater.getIntValue() = 0 and not astNode.isInclusive()
      )
    )
  }

  override DataFlow::Node getBaseString() {
    result = indexOf.getReceiver().flow()
  }

  override DataFlow::Node getSubstring() {
    result = indexOf.getArgument(0).flow()
  }

  override boolean getPolarity() {
    result = polarity
  }
}

/**
 * An expression of form `~A.indexOf(B)` which, when coerced to a boolean, is equivalent to `A.includes(B)`.
 */
private class IndexOfBitwiseContainsCheck extends StringContainsCheck, DataFlow::ValueNode {
  MethodCallExpr indexOf;
  override BitNotExpr astNode;

  IndexOfBitwiseContainsCheck() {
    astNode.getOperand() = indexOf and
    indexOf.getMethodName() = "indexOf"
  }

  override DataFlow::Node getBaseString() {
    result = indexOf.getReceiver().flow()
  }

  override DataFlow::Node getSubstring() {
    result = indexOf.getArgument(0).flow()
  }
}
