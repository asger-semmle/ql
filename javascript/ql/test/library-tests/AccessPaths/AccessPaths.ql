import javascript
import semmle.javascript.dataflow.internal.AccessPaths

from AccessPath path
select path, concat(DataFlow::Node node | node = path.getAnInstance() | node.toString(), ", ")
