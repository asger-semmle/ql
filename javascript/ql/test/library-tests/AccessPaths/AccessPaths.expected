| a | a |
| b | b |
| c | c, c |
| c.z | c.z |
| this |  |
| x | x, x, x, x, x |
| x.y | x.y, x.y, x.y |
| x.y.z | x.y.z, x.y.z |
| y | y, y |
| y.z | y.z |
| z | z |
